Rename modeling exercise server routes to follow REST guidelines

maxr96 opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] ~Client: I added multiple screenshots/screencasts of my UI changes~
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
We want to be consistent in naming our REST Endpoints. The changed endpoints didn't follow our REST Guidelines.

### Description
<!-- Describe your changes in detail -->
The changes reflects the table of REST routes https://confluence.ase.in.tum.de/pages/viewpage.action?pageId=30016376
Changed routes in this PR:
- GET modeling-editor/{participationId} -> participations/{participationId}/latest-submission
- PUT modeling-submissions/{submissionId}/feedback -> modelling-submissions/{submissionId}/assessment
- PUT exercise/{exerciseId}/submission/{submissionId}/modelingExampleAssessment -> modeling-submissions/{submissionId}/example-assessment
- GET modeling-submissions/{submissionId}/result -> modeling-submissions/{submissionId}/assessment
- PUT modeling-exercises -> modeling-exercises/{exerciseId}

**Test Coverage**
Thanks to @madwau all the REST endpoints (Modeling assessment, submission and exercise)  are covered with tests, I adjusted the tests to target renamed endpoints and further improved the test coverage.
![Screenshot_2020-03-05 de tum in www1 artemis web rest](https://user-images.githubusercontent.com/17262376/75931622-bc350f80-5e75-11ea-9f7a-93c4edffb3c5.png)
![Screenshot_2020-03-05 de tum in www1 artemis web rest(1)](https://user-images.githubusercontent.com/17262376/75931630-c1925a00-5e75-11ea-8bcb-019731bbfaa5.png)
![Screenshot_2020-03-05 de tum in www1 artemis web rest(2)](https://user-images.githubusercontent.com/17262376/75932191-27cbac80-5e77-11ea-81da-df26d89a527e.png)

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Course Administration
3. Create and start modeling exercise with example assessment
4. Check that as an instructor you can update it.
5. Participate and submit a solution as a student.
6. Assess this exercise as a tutor.

krusche commented:
Still a draft and not much progress since 17 days. You can reopen it, if the changes are finished!

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java:
```diff
@@ -116,17 +116,18 @@ public ModelingExerciseResource(ModelingExerciseRepository modelingExerciseRepos
      * @throws URISyntaxException if the Location URI syntax is incorrect
      */
-    @PutMapping("/modeling-exercises")
+    @PutMapping("/modeling-exercises/{exerciseId}")
```
Why is the entity id a path variable here? As far as I can see it's normally not included explicitly in the path for our PUT endpoints. Is there a specific reason why it's needed here?

maxr96 replied:
The main reason is to conform to the standards defined for REST endpoints. `PUT` endpoints should always include the entity id. I did couple of changes to make it clear that we need it in the code. Now we also check if the modeling exercise exists in the database prior to storing update modeling exercise

madwau replied:
Oh but then the behavior of the endpoint has been changed by the refactoring since it could be used before for both creating and updating whereas now it can only be used for updating. Did you check if all current usages are still valid with the new behavior?

maxr96 replied:
Yes, I checked that the creation and updating of the modeling exercise works properly. I also decided to change the update for the modeling submission. This endpoint is a bit more dangerous so I will still have to test it.

maxr96 replied:
**UPDATE**
Finished with the testing. Can you check it now?

krusche replied:
This makes our whole REST API inconsistent. I think we should revert this change!

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java:
```diff
@@ -323,9 +323,9 @@ else if (!authCheckService.isAtLeastInstructorForExercise(exercise)) {
      * @return the ResponseEntity with the submission as body
      */
-    @GetMapping("/modeling-editor/{participationId}")
+    @GetMapping("/participations/{participationId}/latest-submission")
```
The naming seems very generic. Shouldn't `modeling` be in it somewhere since it only supports modeling exercises?

maxr96 replied:
True. Missed it somehow. I renamed it to `latest-modeling-submission`

madwau commented on src/test/java/de/tum/in/www1/artemis/ModelingExerciseIntegrationTest.java:
```diff
@@ -117,15 +117,16 @@ public void testCreateModelingExercise_asInstructor() throws Exception {
     public void testUpdateModelingExercise_asInstructor() throws Exception {
         ModelingExercise modelingExercise = modelingExerciseUtilService.createModelingExercise(classExercise.getCourse().getId());
         // The PUT request basically forwards to POST in case the modeling exercise id is not yet set.
```
The comment is no longer relevant.

maxr96 replied:
Good catch!

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java:
```diff
@@ -116,25 +111,28 @@ public ModelingExerciseResource(ModelingExerciseRepository modelingExerciseRepos
         }
 
+        // Check that the modeling exercise id is correct
+        if (!modelingExerciseRepository.existsById(exerciseId) || !exerciseId.equals(modelingExercise.getId())) {
```
I think the second condition should have a different error message since that could be pretty confusing otherwise if the exercise actually exists but there is just an id mismatch.

maxr96 replied:
You are right I split it into two checks. First is done on the `ModelingExerciseService` and second one here. The second one now says `The updated modeling exercise doesn't match the given id` if failed.

maxr96 commented:
> @maxr96 please note: due to the recent critical bugs and problems in your changes, we need full test coverage of all changed files in the server and in the client, otherwise we need to decline this PR. Please have a look into the coverage reports and make sure all changed files are fully covered in the tests. If this is not the case, please add new tests to cover everything!
> 
> Please document this by uploading a screenshot of the coverage report to the PR.

I added test coverage reports for the files I changed. The lines that I changed in this PR are covered with the existing tests.

krusche commented:
please fix the merge conflicts and make sure to further increase the test coverage, also the one for the changed client files!

madwau commented:
@maxr96 Can you resolve the merge conflicts here?

maxr96 commented on src/main/java/de/tum/in/www1/artemis/service/ModelingExerciseService.java:
```diff
@@ -29,4 +29,10 @@ public ModelingExercise findOne(Long exerciseId) {
         return modelingExerciseRepository.findById(exerciseId).orElseThrow(() -> new EntityNotFoundException("Exercise with id: \"" + exerciseId + "\" does not exist"));
     }
+
+    public void existsById(long exerciseId) {
```
Applied

maxr96 commented on src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java:
```diff
@@ -386,4 +386,10 @@ public long countSubmissionsToAssessByCourseId(Long courseId) {
         return modelingSubmissionRepository.countByExerciseIdSubmittedBeforeDueDate(exerciseId);
     }
+
+    public void existsById(long submissionId) {
```
Thank you for the suggestion. This definitely makes more sense

maxr96 commented:
Thank you for checking my PR @madwau and @ungaralex . I resolved the merge conflicts and applied the comments. I also increased the test coverage for `ModelingSubmissionResource` from 93% to 97% and `ModelingExerciseResource` from 93% to 96%. Can you check the PR again please?

krusche commented:
@maxr96 this is still not 100%. Again, we only invest time in this PR, if you can make sure that really *ALL* functionality related to your changes is fully tested on the client and server!

maxr96 commented:
@krusche This line is not reachable as the results of participations are never initialized. That is why the code coverage of `ModelingAssessmentResource` is not 100%. However, for other two resources the coverage is now **exactly** 100%.
![Screenshot_2020-03-05 ModelingAssessmentResource java](https://user-images.githubusercontent.com/17262376/75931794-17670200-5e76-11ea-9cdb-2cd123374ed7.png)

maxr96 commented:
> Hi Maksym, thanks for your contribution. I experienced the following unexpected behavior:
> 
>     * I am not able to create an example solution for a modeling exercise and get the following console error
>       (same with example submission)
>       ![image](https://user-images.githubusercontent.com/50635005/76664125-91d50780-6583-11ea-8d86-ee5ec12cc959.png)
>       Could you please fix this? Thanks.

Hi @HanyaElhashemy , thank you for checking my PR and reporting the issue. I checked that my changes did not cause this bug, as on the latest version of development branch I experience the same issue. However, I will investigate it and check how it can be fixed.

maxr96 commented:
**UPDATE**: I managed to solve the issue with modeling example solution and submission. Turns out that `ArtemisModelingExerciseManagementModule` was simply not imported into the `CourseManagementModule`. I also fixed back button routes for this components. Can you please chek my PR again @HanyaElhashemy ?

madwau commented on src/main/webapp/app/exercises/modeling/manage/modeling-exercise.service.ts:
```diff
@@ -27,11 +27,11 @@ export class ModelingExerciseService {
     }
 
-    update(modelingExercise: ModelingExercise, req?: any): Observable<EntityResponseType> {
+    update(modelingExercise: ModelingExercise, exerciseId: number, req?: any): Observable<EntityResponseType> {
```
Isn't the `exerciseId` argument redundant here since it can be retrieved from `modelingExercise`?

krusche replied:
agreed

maxr96 replied:
Ok, I reverted this change

madwau commented on src/main/webapp/app/exercises/modeling/participate/modeling-submission.service.ts:
```diff
@@ -24,10 +24,10 @@ export class ModelingSubmissionService {
     }
 
-    update(modelingSubmission: ModelingSubmission, exerciseId: number): Observable<EntityResponseType> {
+    update(modelingSubmission: ModelingSubmission, exerciseId: number, submissionId: number): Observable<EntityResponseType> {
```
Same here. Isn't the `submissionId` argument redundant here since it can be retrieved from `modelingSubmission`?

krusche replied:
I agree, let's revert this change!

maxr96 replied:
Ok, I reverted this change

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java:
```diff
@@ -105,24 +105,29 @@ public ModelingSubmissionResource(ModelingSubmissionService modelingSubmissionSe
      *         with status 500 (Internal Server Error) if the modelingSubmission couldn't be updated
      */
-    @PutMapping("/exercises/{exerciseId}/modeling-submissions")
+    @PutMapping("/exercises/{exerciseId}/modeling-submissions/{submissionId}")
```
This makes our whole REST API inconsistent. I think we should revert this change!

maxr96 replied:
Reverted

maxr96 commented:
As requested I removed the ids in `ModelingSubmission` and `ModelingExercise` resources PUT endpoints
